// vidstack/dist/prod/providers/video/presentation/video-presentation.js
import { listenEvent, DOMEvent } from 'maverick.js/std';
import { $ as canUseVideoPresentation } from '../../../media-core.js';

class VideoPresentation {
  constructor(_video, _media) {
    this.aa = _video;
    this.k = _media;
    listenEvent(this.aa, "webkitpresentationmodechanged", this.wi.bind(this));
  }
  Ga = "inline";
  get vi() {
    return canUseVideoPresentation(this.aa);
  }
  async ui(mode) {
    if (this.Ga === mode)
      return;
    this.aa.webkitSetPresentationMode(mode);
  }
  wi() {
    const prevMode = this.Ga;
    this.Ga = this.aa.webkitPresentationMode;
    this.k.player?.dispatchEvent(
      new DOMEvent("video-presentation-change", {
        detail: this.Ga,
        trigger: event
      })
    );
    ["fullscreen", "picture-in-picture"].forEach((type) => {
      if (this.Ga === type || prevMode === type) {
        this.k.delegate.R(`${type}-change`, {
          detail: this.Ga === type,
          trigger: event
        });
      }
    });
  }
}
class FullscreenPresentationAdapter {
  constructor(_presentation) {
    this.ti = _presentation;
  }
  get active() {
    return this.ti.Ga === "fullscreen";
  }
  get supported() {
    return this.ti.vi;
  }
  async enter() {
    this.ti.ui("fullscreen");
  }
  async exit() {
    this.ti.ui("inline");
  }
}
class PIPPresentationAdapter {
  constructor(_presentation) {
    this.ti = _presentation;
  }
  get active() {
    return this.ti.Ga === "picture-in-picture";
  }
  get supported() {
    return this.ti.vi;
  }
  async enter() {
    this.ti.ui("picture-in-picture");
  }
  async exit() {
    this.ti.ui("inline");
  }
}

export { FullscreenPresentationAdapter as F, PIPPresentationAdapter as P, VideoPresentation as V };
